homeassistant/components/http.py

```diff
@@ -152,105 +164,84 @@ def stop_wsgi_server(event):
+    """FileSender class capable of sending gzip version if available."""
 
-    return Request
+    # pylint: disable=invalid-name, too-few-public-methods
 
+    development = False
 
-def routing_map(hass):
-    """Generate empty routing map with HA validators."""
-    from werkzeug.routing import Map, BaseConverter, ValidationError
+    @asyncio.coroutine
+    def send(self, request, filepath):
+        """Send filepath to client using request."""
+        gzip = False
+        if 'gzip' in request.headers[hdrs.ACCEPT_ENCODING]:
+            gzip_path = filepath.with_name(filepath.name + '.gz')
 
-    class EntityValidator(BaseConverter):
-        """Validate entity_id in urls."""
+            if gzip_path.is_file():
+                filepath = gzip_path
+                gzip = True
 
-        regex = r"(\w+)\.(\w+)"
+        st = filepath.stat()
 
-        def __init__(self, url_map, exist=True, domain=None):
-            """Initilalize entity validator."""
-            super().__init__(url_map)
-            self._exist = exist
-            self._domain = domain
+        modsince = request.if_modified_since
+        if modsince is not None and st.st_mtime <= modsince.timestamp():
+            raise HTTPNotModified()
 
-        def to_python(self, value):
-            """Validate entity id."""
-            if self._exist and hass.states.get(value) is None:
-                raise ValidationError()
-            if self._domain is not None and \
-               split_entity_id(value)[0] != self._domain:
-                raise ValidationError()
+        ct, encoding = mimetypes.guess_type(str(filepath))
+        if not ct:
+            ct = 'application/octet-stream'
 
-            return value
+        resp = self._response_factory()
+        resp.content_type = ct
+        if encoding:
+            resp.headers[hdrs.CONTENT_ENCODING] = encoding
+        if gzip:
+            resp.headers[hdrs.VARY] = hdrs.ACCEPT_ENCODING
+        resp.last_modified = st.st_mtime
 
-        def to_url(self, value):
-            """Convert entity_id for a url."""
-            return value
+        # CACHE HACK
+        if not self.development:
+            cache_time = 31 * 86400  # = 1 month
+            resp.headers[hdrs.CACHE_CONTROL] = "public, max-age={}".format(
+                cache_time)
 
-    class DateValidator(BaseConverter):
-        """Validate dates in urls."""
+        file_size = st.st_size
 
-        regex = r'\d{4}-\d{1,2}-\d{1,2}'
-
-        def to_python(self, value):
-            """Validate and convert date."""
-            parsed = dt_util.parse_date(value)
-
-            if parsed is None:
-                raise ValidationError()
+        resp.content_length = file_size
+        resp.set_tcp_cork(True)
+        try:
+            with filepath.open('rb') as f:
+                yield from self._sendfile(request, resp, f, file_size)
 
-            return parsed
+        finally:
+            resp.set_tcp_nodelay(True)
```

Comment:
This code is a direct copy of the aiohttp code. I am planning on contributing it back to them. For now this is a direct copy of their code.

`set_tcp_nodelay` will reverse the `set_tcp_cork` call a few lines above. [Docs.](http://aiohttp.readthedocs.io/en/stable/web_reference.html#aiohttp.web.StreamResponse.set_tcp_cork)